sql: broken cleanup with apply join when an error is encountered

[(roachtest).sqlsmith/setup=empty/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2258516&tab=buildLog) on [release-20.2@05dec53e9980c0203324334baaf9b381fd070d39](https://github.com/cockroachdb/cockroach/commits/05dec53e9980c0203324334baaf9b381fd070d39):

```
																							)
																								AS tab_35388
																									(col_60828, col_60829)
																						LIMIT
																							1:::INT8
																					)
																				),
																				('':::STRING)
																		)
																			AS tab_35389 (col_60831)
																)
														SELECT
															'71 years 9 mons 76 days 09:27:11.662639':::INTERVAL AS col_60832
														FROM
															with_11083 AS cte_ref_3326
														ORDER BY
															cte_ref_3326.col_60831 DESC
														LIMIT
															1:::INT8
													)
												),
												('1975-05-05':::DATE, '60 years 9 mons 449 days 02:18:50.743747':::INTERVAL)
										)
											AS tab_35390 (col_60833, col_60834)
								)
						SELECT
							tab_35380.col_60817 AS col_60835
						FROM
							with_11084 AS cte_ref_3327
						ORDER BY
							cte_ref_3327.col_60834 ASC, cte_ref_3327.col_60834 DESC
						LIMIT
							1:::INT8
					)
						AS col_60836
				LIMIT
					1:::INT8
			)
				AS col_60837
		FROM
			(
				VALUES
					(1416123880:::OID, NULL),
					(2321016888:::OID, 3159163068:::OID),
					(155650621:::OID, 2791005432:::OID),
					(390985191:::OID, NULL),
					(1284045715:::OID, COALESCE(630466950:::OID, NULL)),
					(1274674334:::OID, 2682598554:::OID)
			)
				AS tab_35380 (col_60816, col_60817);

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=empty/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2258516&tab=artifacts#/sqlsmith/setup=empty/setting=no-mutations)
Related:
- #53618 roachtest: sqlsmith/setup=empty/setting=no-mutations failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-master](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-master)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dempty%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
sqlsmith.go:172,sqlsmith.go:202,test_runner.go:754: error: pq: internal error: invalid index 1 for "(SELECT tab_35384.col_60822 AS col_60830 FROM (VALUES (e'\\x1d:t\\t<':::STRING), (e'\\x02\\x19Js':::STRING), (e'O\\x1e\\x16!\\x06\\x1ai':::STRING), ('\"Bc':::STRING)) AS tab_35384 (col_60822), (VALUES (NULL)) AS tab_35385 (col_60823), (VALUES (ARRAY[1478050095, 1300076119, 4239289200, 2687116627], (WITH with_11082 (col_60826) AS (SELECT '488a7403-c3fb-42d1-bd9e-94d17a224243':::UUID AS col_60826 FROM (VALUES (e';r\\x07)\\tM0/':::STRING:::NAME)) AS tab_35386 (col_60824), (VALUES (NULL), ('1970-01-10 15:35:16.000549':::TIMESTAMP), ('1970-01-07 14:02:07.000673':::TIMESTAMP), ('1970-01-11 02:40:59.000242':::TIMESTAMP), ('1970-01-04 22:44:48.000992':::TIMESTAMP), ('1970-01-09 12:28:32.000872':::TIMESTAMP)) AS tab_35387 (col_60825)) SELECT 'e113:2913:47c7:3aba:766f:9f26:7c31:d048/118':::INET AS col_60827 FROM with_11082 AS cte_ref_3325 LIMIT 1:::INT8)), (ARRAY[]:::OID[], '155.90.132.56/7':::INET), (ARRAY[385847279, 1978397171, 187263712, 541995921], '7f4f:77a0:3c96:f2d4:aa09:c647:9610:c3ce/30':::INET), (NULL, '65.147.109.104/12':::INET)) AS tab_35388 (col_60828, col_60829) LIMIT 1:::INT8)"
		stmt:
		SELECT
			357289228:::OID AS col_60818,
			(
				SELECT
					(
						WITH
							with_11084 (col_60833, col_60834)
								AS (
									SELECT
										*
									FROM
										(
											VALUES
												('1993-06-05':::DATE, '45 years 8 mons 319 days 05:47:59.241468':::INTERVAL),
												(
													'4714-11-24 BC':::DATE,
													(
														WITH
															with_11080 (col_60819)
																AS (
																	SELECT
																		*
																	FROM
																		(
																			VALUES
																				(
																					ARRAY[(-29547):::INT8,(-28304):::INT8,(-128):::INT8,(-22657):::INT8,(-1):::INT8,(-9665):::INT8]
																				),
																				(
																					ARRAY[(-5766):::INT8,(-30421):::INT8,(-9404):::INT8]
																				),
																				(
																					ARRAY[2862:::INT8,31486:::INT8,14360:::INT8,32663:::INT8,3351:::INT8,(-15108):::INT8]
																				),
																				(
																					ARRAY[(-10467):::INT8,14757:::INT8,29608:::INT8,7196:::INT8]
																				),
																				(NULL)
																		)
																			AS tab_35381 (col_60819)
																	UNION ALL
																		SELECT
																			*
																		FROM
																			(
																				VALUES
																					(ARRAY[(-17372):::INT8,(-20420):::INT8]),
																					(ARRAY[(-26944):::INT8,(-4719):::INT8]),
																					(NULL),
																					(NULL)
																			)
																				AS tab_35382 (col_60820)
																),
															with_11081 (col_60821)
																AS (
																	SELECT
																		*
																	FROM
																		(
																			VALUES
																				('01:46:40.671173+01:46:00':::TIMETZ),
																				('15:57:23.293415+04:02:00':::TIMETZ)
																		)
																			AS tab_35383 (col_60821)
																),
															with_11083 (col_60831)
																AS (
																	SELECT
																		*
																	FROM
																		(
																			VALUES
																				(' <:qg!QN':::STRING),
																				('':::STRING),
																				('nC U20':::STRING),
																				(
																					(
																						SELECT
																							tab_35384.col_60822 AS col_60830
																						FROM
																							(
																								VALUES
																									(e'\x1d:t\t<':::STRING),
																									(e'\x02\x19Js':::STRING),
																									(
																										e'O\x1e\x16!\x06\x1ai':::STRING
																									),
																									('"Bc':::STRING)
																							)
																								AS tab_35384 (col_60822),
																							(VALUES (NULL))
																								AS tab_35385 (col_60823),
																							(
																								VALUES
																									(
																										ARRAY[1478050095:::OID,1300076119:::OID,4239289200:::OID,2687116627:::OID],
																										(
																											WITH
																												with_11082
																													(col_60826)
																													AS (
																														SELECT
																															'488a7403-c3fb-42d1-bd9e-94d17a224243':::UUID
																																AS col_60826
																														FROM
																															(
																																VALUES
																																	(
																																		e';r\x07)\tM0/':::STRING:::NAME
																																	)
																															)
																																AS tab_35386
																																	(
																																		col_60824
																																	),
																															(
																																VALUES
																																	(
																																		NULL
																																	),
																																	(
																																		'1970-01-10 15:35:16.000549':::TIMESTAMP
																																	),
																																	(
																																		'1970-01-07 14:02:07.000673':::TIMESTAMP
																																	),
																																	(
																																		'1970-01-11 02:40:59.000242':::TIMESTAMP
																																	),
																																	(
																																		'1970-01-04 22:44:48.000992':::TIMESTAMP
																																	),
																																	(
																																		'1970-01-09 12:28:32.000872':::TIMESTAMP
																																	)
																															)
																																AS tab_35387
																																	(
																																		col_60825
																																	)
																													)
																											SELECT
																												'e113:2913:47c7:3aba:766f:9f26:7c31:d048/118':::INET
																													AS col_60827
																											FROM
																												with_11082
																													AS cte_ref_3325
																											LIMIT
																												1:::INT8
																										)
																									),
																									(
																										ARRAY[]:::OID[],
																										'155.90.132.56/7':::INET
																									),
																									(
																										ARRAY[385847279:::OID,1978397171:::OID,187263712:::OID,541995921:::OID],
																										'7f4f:77a0:3c96:f2d4:aa09:c647:9610:c3ce/30':::INET
																									),
																									(
																										NULL,
																										'65.147.109.104/12':::INET
																									)
																							)
																								AS tab_35388
																									(col_60828, col_60829)
																						LIMIT
																							1:::INT8
																					)
																				),
																				('':::STRING)
																		)
																			AS tab_35389 (col_60831)
																)
														SELECT
															'71 years 9 mons 76 days 09:27:11.662639':::INTERVAL AS col_60832
														FROM
															with_11083 AS cte_ref_3326
														ORDER BY
															cte_ref_3326.col_60831 DESC
														LIMIT
															1:::INT8
													)
												),
												('1975-05-05':::DATE, '60 years 9 mons 449 days 02:18:50.743747':::INTERVAL)
										)
											AS tab_35390 (col_60833, col_60834)
								)
						SELECT
							tab_35380.col_60817 AS col_60835
						FROM
							with_11084 AS cte_ref_3327
						ORDER BY
							cte_ref_3327.col_60834 ASC, cte_ref_3327.col_60834 DESC
						LIMIT
							1:::INT8
					)
						AS col_60836
				LIMIT
					1:::INT8
			)
				AS col_60837
		FROM
			(
				VALUES
					(1416123880:::OID, NULL),
					(2321016888:::OID, 3159163068:::OID),
					(155650621:::OID, 2791005432:::OID),
					(390985191:::OID, NULL),
					(1284045715:::OID, COALESCE(630466950:::OID, NULL)),
					(1274674334:::OID, 2682598554:::OID)
			)
				AS tab_35380 (col_60816, col_60817);
```

```
SELECT
			357289228:::OID AS col_60818,
			(
				SELECT
					(
						WITH
							with_11084 (col_60833, col_60834)
								AS (
									SELECT
										*
									FROM
										(
											VALUES
												('1993-06-05':::DATE, '45 years 8 mons 319 days 05:47:59.241468':::INTERVAL),
												(
													'4714-11-24 BC':::DATE,
													(
														WITH
															with_11080 (col_60819)
																AS (
																	SELECT
																		*
																	FROM
																		(
																			VALUES
																				(
																					ARRAY[(-29547):::INT8,(-28304):::INT8,(-128):::INT8,(-22657):::INT8,(-1):::INT8,(-9665):::INT8]
																				),
																				(
																					ARRAY[(-5766):::INT8,(-30421):::INT8,(-9404):::INT8]
																				),
																				(
																					ARRAY[2862:::INT8,31486:::INT8,14360:::INT8,32663:::INT8,3351:::INT8,(-15108):::INT8]
																				),
																				(
																					ARRAY[(-10467):::INT8,14757:::INT8,29608:::INT8,7196:::INT8]
																				),
																				(NULL)
																		)
																			AS tab_35381 (col_60819)
																	UNION ALL
																		SELECT
																			*
																		FROM
																			(
																				VALUES
																					(ARRAY[(-17372):::INT8,(-20420):::INT8]),
																					(ARRAY[(-26944):::INT8,(-4719):::INT8]),
																					(NULL),
																					(NULL)
																			)
																				AS tab_35382 (col_60820)
																),
															with_11081 (col_60821)
																AS (
																	SELECT
																		*
																	FROM
																		(
																			VALUES
																				('01:46:40.671173+01:46:00':::TIMETZ),
																				('15:57:23.293415+04:02:00':::TIMETZ)
																		)
																			AS tab_35383 (col_60821)
																),
															with_11083 (col_60831)
																AS (
																	SELECT
																		*
																	FROM
																		(
																			VALUES
																				(' <:qg!QN':::STRING),
																				('':::STRING),
																				('nC U20':::STRING),
																				(
																					(
																						SELECT
																							tab_35384.col_60822 AS col_60830
																						FROM
																							(
																								VALUES
																									(e'\x1d:t\t<':::STRING),
																									(e'\x02\x19Js':::STRING),
																									(
																										e'O\x1e\x16!\x06\x1ai':::STRING
																									),
																									('"Bc':::STRING)
																							)
																								AS tab_35384 (col_60822),
																							(VALUES (NULL))
																								AS tab_35385 (col_60823),
																							(
																								VALUES
																									(
																										ARRAY[1478050095:::OID,1300076119:::OID,4239289200:::OID,2687116627:::OID],
																										(
																											WITH
																												with_11082
																													(col_60826)
																													AS (
																														SELECT
																															'488a7403-c3fb-42d1-bd9e-94d17a224243':::UUID
																																AS col_60826
																														FROM
																															(
																																VALUES
																																	(
																																		e';r\x07)\tM0/':::STRING:::NAME
																																	)
																															)
																																AS tab_35386
																																	(
																																		col_60824
																																	),
																															(
																																VALUES
																																	(
																																		NULL
																																	),
																																	(
																																		'1970-01-10 15:35:16.000549':::TIMESTAMP
																																	),
																																	(
																																		'1970-01-07 14:02:07.000673':::TIMESTAMP
																																	),
																																	(
																																		'1970-01-11 02:40:59.000242':::TIMESTAMP
																																	),
																																	(
																																		'1970-01-04 22:44:48.000992':::TIMESTAMP
																																	),
																																	(
																																		'1970-01-09 12:28:32.000872':::TIMESTAMP
																																	)
																															)
																																AS tab_35387
																																	(
																																		col_60825
																																	)
																													)
																											SELECT
																												'e113:2913:47c7:3aba:766f:9f26:7c31:d048/118':::INET
																													AS col_60827
																											FROM
																												with_11082
																													AS cte_ref_3325
																											LIMIT
																												1:::INT8
																										)
																									),
																									(
																										ARRAY[]:::OID[],
																										'155.90.132.56/7':::INET
																									),
																									(
																										ARRAY[385847279:::OID,1978397171:::OID,187263712:::OID,541995921:::OID],
																										'7f4f:77a0:3c96:f2d4:aa09:c647:9610:c3ce/30':::INET
																									),
																									(
																										NULL,
																										'65.147.109.104/12':::INET
																									)
																							)
																								AS tab_35388
																									(col_60828, col_60829)
																						LIMIT
																							1:::INT8
																					)
																				),
																				('':::STRING)
																		)
																			AS tab_35389 (col_60831)
																)
														SELECT
															'71 years 9 mons 76 days 09:27:11.662639':::INTERVAL AS col_60832
														FROM
															with_11083 AS cte_ref_3326
														ORDER BY
															cte_ref_3326.col_60831 DESC
														LIMIT
															1:::INT8
													)
												),
												('1975-05-05':::DATE, '60 years 9 mons 449 days 02:18:50.743747':::INTERVAL)
										)
											AS tab_35390 (col_60833, col_60834)
								)
						SELECT
							tab_35380.col_60817 AS col_60835
						FROM
							with_11084 AS cte_ref_3327
						ORDER BY
							cte_ref_3327.col_60834 ASC, cte_ref_3327.col_60834 DESC
						LIMIT
							1:::INT8
					)
						AS col_60836
				LIMIT
					1:::INT8
			)
				AS col_60837
		FROM
			(
				VALUES
					(1416123880:::OID, NULL),
					(2321016888:::OID, 3159163068:::OID),
					(155650621:::OID, 2791005432:::OID),
					(390985191:::OID, NULL),
					(1284045715:::OID, COALESCE(630466950:::OID, NULL)),
					(1274674334:::OID, 2682598554:::OID)
			)
				AS tab_35380 (col_60816, col_60817);

*
*
* ERROR: [n1,client=127.0.0.1:58607,hostssl,user=root] a panic has occurred!
* txn: unexpected 10240 leftover bytes
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:508
*   | runtime.gopanic
*   |   /usr/local/Cellar/go@1.13/1.13.15/libexec/src/runtime/panic.go:679
*   | github.com/cockroachdb/cockroach/pkg/util/log.ReportOrPanic
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:340
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:404
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:391
*   | github.com/cockroachdb/cockroach/pkg/sql.(*txnState).finishSQLTxn
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/txn_state.go:229
*   | github.com/cockroachdb/cockroach/pkg/sql.cleanupAndFinishOnError
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_fsm.go:475
*   | github.com/cockroachdb/cockroach/pkg/util/fsm.Transitions.apply
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:106
*   | github.com/cockroachdb/cockroach/pkg/util/fsm.(*Machine).ApplyWithPayload
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:135
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).txnStateTransitionsApplyWrapper
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2208
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1582
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1388
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:510
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:626
* Wraps: (2) while executing: SELECT _:::OID AS _, (SELECT (WITH _ (_, _) AS (SELECT * FROM (VALUES (_:::DATE, _:::INTERVAL), (__more2__)) AS _ (_, _)) SELECT _._ AS _ FROM _ AS _ ORDER BY _._ ASC, _._ DESC LIMIT _:::INT8) AS _ LIMIT _:::INT8) AS _ FROM (VALUES (_:::OID, _), (__more5__)) AS _ (_, _)
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:508
*   | runtime.gopanic
*   |   /usr/local/Cellar/go@1.13/1.13.15/libexec/src/runtime/panic.go:679
*   | [...repeated from below...]
* Wraps: (4) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/util/log.ReportOrPanic
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/log/crash_reporting.go:338
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).doStop
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:404
*   | github.com/cockroachdb/cockroach/pkg/util/mon.(*BytesMonitor).Stop
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:391
*   | github.com/cockroachdb/cockroach/pkg/sql.(*txnState).finishSQLTxn
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/txn_state.go:229
*   | github.com/cockroachdb/cockroach/pkg/sql.cleanupAndFinishOnError
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_fsm.go:475
*   | github.com/cockroachdb/cockroach/pkg/util/fsm.Transitions.apply
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:106
*   | github.com/cockroachdb/cockroach/pkg/util/fsm.(*Machine).ApplyWithPayload
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/util/fsm/fsm.go:135
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).txnStateTransitionsApplyWrapper
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2208
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1582
*   | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1388
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:510
*   | github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1
*   |   /Users/jordan/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:626
*   | runtime.goexit
*   |   /usr/local/Cellar/go@1.13/1.13.15/libexec/src/runtime/asm_amd64.s:1357
* Wraps: (5) txn: unexpected 10240 leftover bytes
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) *withstack.withStack (5) *errutil.leafError
*
panic: txn: unexpected 10240 leftover bytes [recovered]
        panic: txn: unexpected 10240 leftover bytes
```

--- a/pkg/sql/apply_join.go
+++ b/pkg/sql/apply_join.go
@@ -249,7 +249,7 @@ func runPlanInsidePlan(
 		params.extendedEvalCtx.copy,
 		plan.subqueryPlans,
 		recv,
-		true,
+		false, /* maybeDistribute */
 	) {
 		if err := rowResultWriter.Err(); err != nil {
 			return err

--- a/pkg/sql/distsql_running.go
+++ b/pkg/sql/distsql_running.go
@@ -806,7 +806,10 @@ func (r *DistSQLReceiver) Types() []*types.T {
 }
 
 // PlanAndRunSubqueries returns false if an error was encountered and sets that
-// error in the provided receiver.
+// error in the provided receiver. Note that if false is returned, then this
+// function will have closed all the subquery plans because it assumes that the
+// caller will not try to run the main plan given that the subqueries'
+// evaluation failed.
 func (dsp *DistSQLPlanner) PlanAndRunSubqueries(
 	ctx context.Context,
 	planner *planner,
@@ -827,6 +830,14 @@ func (dsp *DistSQLPlanner) PlanAndRunSubqueries(
 			maybeDistribute,
 		); err != nil {
 			recv.SetError(err)
+			// Usually we leave the closure of subqueries to occur when the
+			// whole plan is being closed (i.e. planTop.close); however, since
+			// we've encountered an error, we might never get to the point of
+			// closing the whole plan, so we choose to defensively close the
+			// subqueries here.
+			for i := range subqueryPlans {
+				subqueryPlans[i].plan.Close(ctx)
+			}
 			return false
 		}
 	}

--- a/pkg/sql/logictest/testdata/logic_test/apply_join
+++ b/pkg/sql/logictest/testdata/logic_test/apply_join
@@ -351,3 +351,31 @@ UPDATE cpk SET extra = (
     WHERE k='k1'
 )
 WHERE ((cpk.key, cpk.value) IN (SELECT new_values.k, new_values.v FROM new_values))
+
+# Regression test for not closing the subqueries in the apply join if they hit
+# an error (#54166).
+query error pgcode XX000 invalid index .*
+SELECT
+  (
+    SELECT
+      tab_4.col_4
+    FROM
+      (VALUES (1)) AS tab_1 (col_1)
+      JOIN (
+          VALUES
+            (
+              (
+                SELECT
+                  1
+                FROM
+                  (SELECT 1)
+                WHERE
+                  EXISTS(SELECT 1)
+              )
+            )
+        )
+          AS tab_6 (col_6) ON (tab_1.col_1) = (tab_6.col_6)
+  )
+FROM
+  (VALUES (NULL)) AS tab_4 (col_4),
+  (VALUES (NULL), (NULL)) AS tab_5 (col_5)

